vm/src/builtins/object.rs

```diff
@@ -184,8 +185,23 @@ impl PyBaseObject {
 
     /// Return repr(self).
     #[pymethod(magic)]
-    fn repr(zelf: PyObjectRef) -> String {
-        format!("<{} object at {:#x}>", zelf.class().name, zelf.get_id())
+    fn repr(zelf: PyObjectRef, vm: &VirtualMachine) -> String {
+        let class = zelf.class();
+        match class.module(vm).downcast_ref::<PyStr>().map(|m| m.as_str()) {
+            Some(module) if module != "builtins" => {
+                format!(
+                    "<{}.{} object at {:#x}>",
+                    module,
+                    class
+                        .qualname(vm)
+                        .downcast_ref::<PyStr>()
+                        .map(|n| n.as_str())
+                        .unwrap_or_else(|| class.name.as_str()),
+                    zelf.get_id()
```

Comment:
is this also different from `PyType::tp_name()`? Maybe because builtins module.

Reply:
I'm so sorry but I couldn't understand your comment in sure in my English skill. what points are different between those lines and `PyType::tp_name`? Do you mean that `PyType::tp_name` is also different from CPython `tp_name`?

Reply:
Sorry. I just ran test if it is worth to discuss or not.

`class.tp_name()` is not just name(`class.name()`) but `{module}.{name}` by itself. So If the fallback case is using `class.tp_name()`, I think we don't need to use match expression here.

Reply:
Ah, perhaps, I understood well. Actually I suspected the `PyType::tp_name()`.

While researching this issue, I tried to change `format!("<{} object at {:#x}>", zelf.name.clone(), zelf.get_id())` into `format!("<{} object at {:#x}>", class.tp_name(), zelf.get_id())` but the `PyType::tp_name()` hasn't returned `{module}.{qualname}` (i.e. `{module}.{full_path}`) but `{module}.{class_name}`.

For instance, if there are codes like below and I changed `format!("<{} object at {:#x}>", zelf.name.clone(), zelf.get_id())` into `format!("<{} object at {:#x}>", class.tp_name(), zelf.get_id())`:

```python
class F:
    def f(self):
        class A: ...

        print(repr(A()))

F().f()
```

We expected `<__main__.F.f.<locals>.A object at 0x7fed36f58790>` but `<__main__.A object at 0x55f0665da830>`.

So I tried to correct `PyType::tp_name()` but I confused while comparing with CPython because CPython implementation stores `tp_name` and derives `name` from the `tp_name` while RustPython derives `tp_name` from `name`. Is it intended? I don't have opinion which it is bad but I wonder the reason 🤔

Reply:
I didn't investigate CPython code yet, but I am pretty sure `{module}.{class_name}` is a bug. If that was the reason, I think this code must simply use `tp_name`, and then we also need to fix tp_name later for compatibility. How do you think about it?

The difference between RustPython and CPython tp_name is not made by design. So it is not intended.

Reply:
I understood and agree with your opinion that this code must use `tp_name` and we should fix `tp_name`.

I will remove all commits and add a commit to replace `zelf.class().name` with `zelf.class().tp_name()`. And I will continue this pull request with direction to fix the `tp_name`.

Reply:
Hou about finishing this PR by replacing it to tp_name and fix tp_name later in different patch?

Reply:
Ok, it seems better 👌